### PR TITLE
FFI callbacks and vtables

### PR DESCRIPTION
**Pull request by bendk**
This PR updates the callback interface code by adding new `FfiType` variants:

- `FfiType::Callback` is a callback function.
- `FfiType::Struct` is a repr(C) struct of `FfiType` fields.
- `FfiType::Reference` is a pointer/reference to another `FfiType`.

`FfiType::Callback` and `FfiType::Struct` store a name that corresponds to a callback/struct definition that the foreign bindings render.  IOW, we can now define callbacks/structs in the `ComponentInterface`, then refer to them by name as ffi types.

These types are used to replace the callback interface code.  Instead of having the foreign code register a  `FfiType::ForeignCallback`, it now registers a reference to a vtable: a struct where each field is a callback that implements one of the callback interface methods, plus a field for the `uniffi_free` method.

The primary motivation for this is to prepare async callback methods.  I think those methods should input an extra argument that's a callback to call when the async method completes.  However, that presents a couple challenges to the current system:
- That callback would be generic on the `FfiType`, which is a new thing that could get very messy.  This code will allow us to handle the monomorphization in the `ComponentInterface` code and keep the bindings generators simple.
- Currently all arguments are passed in a `RustBuffer`, but I don't want to read a function pointer out of that.  For one, it feels like too much transmuting to me and I don't like making too many assumptions about function pointers.  Secondly, it seems like it's mixing abstractions.  The `read` operation should be used for `Type` values, not for `FfiType`.  VTables give us the flexibility to avoid this since we can just add a new `FfiType` parameter for async methods only.

I think this change will also improve our lives in general.  It's currently pretty tedious to add a new callback type or refactor an existing one.  I think this system will make that easier with a lot less fiddling around to get the signatures and FFI types correct for each language.  Also, the concept of vtables opens up some interesting possibilities and if we had that before the initial async work it could have simplified some things.

### CODE REVIEW

File: fixtures/keywords/kotlin/src/lib.rs [hunk 1]
```diff
@@ -15,9 +15,9 @@ impl r#break {
 }
 
 #[allow(non_camel_case_types)]
-trait r#continue {
+trait r#continue: Send + Sync {
     fn r#return(&self, v: r#return) -> r#return;
-    fn r#continue(&self) -> Option<Box<dyn r#continue>>;
+    fn r#continue(&self) -> Option<Arc<dyn r#continue>>;
```
Comment (mhammond) [inline fixtures/keywords/kotlin/src/lib.rs:20]: I don't understand why these changes are necessary - it implies something might have changed that we should document, but I'm really not clear why this is the only fixture impacted!

**Review by mhammond — APPROVED** [review]
This is a large patch, but it makes a lof of sense and makes callbacks more first-class instead of bolted on. I think you need a CHANGELOG entry for the maintainers of other bindings, and we should avoid landing this until 0.26 is (nearly?) out the door, but it looks great, thanks!

File: uniffi_bindgen/src/interface/ffi.rs [hunk 2]
```diff
@@ -47,15 +47,25 @@ pub enum FfiType {
...
     /// The provider of this reference must keep it alive for the duration of the receiving call.
     ForeignBytes,
-    /// Pointer to a callback function that handles all callbacks on the foreign language side.
-    ForeignCallback,
+    /// Pointer to a callback function.  The inner value which matches one of the items in
+    /// [crate::ComponentInterface::ffi_callback_definitions].
+    Callback(String),
+    /// Pointer to a VTable.  The inner value matches one of the items in
+    /// [crate::ComponentInterface::ffi_struct_definitions].
+    Struct(String),
```
[the remark this answers is not in the document]
Comment (bendk) [reply uniffi_bindgen/src/interface/ffi.rs:55]: Yup that's completely right.  I do have some other ideas for `FfiStructs` but as of now VTables are the only one.

File: uniffi_macros/src/export/callback_interface.rs [hunk 3]
```diff
@@ -8,41 +8,68 @@ use crate::{
...
+        .iter()
+        .map(|sig| gen_method_impl(sig, &vtable_cell))
+        .collect::<syn::Result<Vec<_>>>()?;
+
     Ok(quote! {
-        #[doc(hidden)]
-        static #internals_ident: ::uniffi::ForeignCallbackInternals = ::uniffi::ForeignCallbackInternals::new();
+        struct #vtable_type {
+            #(#vtable_fields)*
+            uniffi_dec_ref: extern "C" fn(handle: u64),
```
[the remark this answers is not in the document]
Comment (bendk) [reply uniffi_macros/src/export/callback_interface.rs:65]: `dec_ref` takes the place of drop now.

File: uniffi_bindgen/src/bindings/ruby/gen_ruby/mod.rs [hunk 4]
```diff
@@ -152,12 +152,15 @@ mod filters {
...
-            // Callback interfaces are not yet implemented, but this needs to return something in
-            // order for the coverall tests to pass.
-            FfiType::ForeignCallback => ":pointer".to_string(),
-            FfiType::RustFutureHandle
-            | FfiType::RustFutureContinuationCallback
-            | FfiType::RustFutureContinuationData => {
+            FfiType::Callback(_) => unimplemented!("FFI Callbacks not implemented"),
+            // Note: references are not really used by the Ruby bindings yet, maybe this should
+            // change to a different type when they are actually implemented.
+            FfiType::Reference(_) => ":pointer".to_string(),
```
[the remark this answers is not in the document]
Comment (bendk) [reply uniffi_bindgen/src/bindings/ruby/gen_ruby/mod.rs:160]: Unfortunately, the tests fail if it's unimplemented and this is the easiest way to make them pass.  I updated the docstring to explain the situation better.

File: uniffi_macros/src/export/callback_interface.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bendk) [reply uniffi_macros/src/export/callback_interface.rs:65]: Oops, I thought you were suggesting to add a `uniffi_drop` field, but now I realize you were probably just saying that the name doesn't make sense which is true.  I'll rename this to `uniffi_drop`.

[the remark this answers is not in the document]
Comment (mhammond) [reply uniffi_macros/src/export/callback_interface.rs:65]: less "doesn't make sense" and more "would be more consistent and clearer" but yeah, it was all about the name.

[the remark this answers is not in the document]
Comment (bendk) [reply uniffi_macros/src/export/callback_interface.rs]: I went with `uniffi_free`, which matches the name in the foreign code and `uniffi_core`.

**Comment by bendk** [comment]
I think this one is ready now, officially re-requesting review.

**Review by mhammond — APPROVED** [review]
🚢

File: uniffi_bindgen/src/bindings/kotlin/templates/ObjectTemplate.kt [hunk 5]
```diff
@@ -304,8 +304,8 @@ open class {{ impl_class_name }}: Disposable, AutoCloseable, {{ interface_name }
 }
 
 {%- if obj.has_callback_interface() %}
-{%- let callback_handler_class = format!("UniffiCallbackInterface{}", name) %}
-{%- let callback_handler_obj = format!("uniffiCallbackInterface{}", name) %}
+{%- let vtable = obj.vtable().expect("trait interface should have a vtable") %}
```
[the remark this answers is not in the document]
Comment (bendk) [reply uniffi_bindgen/src/bindings/kotlin/templates/ObjectTemplate.kt:307]: It depends on `obj` though, so I don't think I can move it.  I think of this as `ObjectTemplate.kt` passing arguments to `CallbalkInterfaceImplTemplate.kt`
